File: Hard/2026-08-03-1406-Stone-Game-III/solution.js
var stoneGameIII = function(stoneValue) {
    const n = stoneValue.length;
    const dp = new Array(n + 1).fill(-Infinity);
    dp[n] = 0;

    for (let i = n - 1; i >= 0; i--) {
        let sum = 0;
        for (let j = 1; j <= 3; j++) {
            if (i + j - 1 < n) {
                sum += stoneValue[i + j - 1];
                dp[i] = Math.max(dp[i], sum - dp[i + j]);
            }
        }
    }

    if (dp[0] > 0) return "Alice";
    if (dp[0] < 0) return "Bob";
    return "Tie";
};

File: Hard/2026-08-03-1406-Stone-Game-III/solution.java
class Solution {
    public String stoneGameIII(int[] stoneValue) {
        int n = stoneValue.length;
        int[] dp = new int[n + 1];
        
        for (int i = n - 1; i >= 0; i--) {
            int maxScore = Integer.MIN_VALUE;
            int sum = 0;
            for (int j = 0; j < 3 && i + j < n; j++) {
                sum += stoneValue[i + j];
                maxScore = Math.max(maxScore, sum - dp[i + j + 1]);
            }
            dp[i] = maxScore;
        }
        
        if (dp[0] > 0) return "Alice";
        else if (dp[0] < 0) return "Bob";
        else return "Tie";
    }
}